pkg/mcs/router/server/sync.go

```diff
@@ -0,0 +1,367 @@
+				zap.String("server", s.name), zap.Uint64("next-index", s.nextSyncIndex))
+		case <-time.After(retryInterval):
+			s.reconnectCh <- false
+		}
+	}()
+	defer logutil.LogPanic()
+	defer s.wg.Done()
+	defer s.streamingRunning.Store(false)
+	// used to load region from kv storage to cache storage.
+	bc := s.cluster
+	// establish client.
+	conn := grpcutil.CreateClientConn(ctx, leaderAddr, s.tlsConfig,
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(msgSize)),
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:    keepaliveTime,
+			Timeout: keepaliveTimeout,
+		}),
+		grpc.WithConnectParams(grpc.ConnectParams{
+			Backoff: backoff.Config{
+				BaseDelay:  time.Second,     // Default was 1s.
+				Multiplier: 1.6,             // Default
+				Jitter:     0.2,             // Default
+				MaxDelay:   3 * time.Second, // Default was 120s.
+			},
+			MinConnectTimeout: 5 * time.Second,
+		}),
+		// WithBlock will block the dial step until success or cancel the context.
+		grpc.WithBlock())
+	// it means the context is canceled.
+	if conn == nil {
+		return
+	}
+	defer conn.Close()
+
+	// Start syncing data.
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		stream, err := s.syncRegion(ctx, conn)
+		failpoint.Inject("disableClientStreaming", func() {
+			err = errors.Errorf("no stream")
+		})
+		if err != nil {
+			if ev, ok := status.FromError(err); ok {
+				if ev.Code() == codes.Canceled {
+					return
+				}
+			}
+			log.Warn("server failed to establish sync stream with leader", zap.String("server", s.name), zap.String("leader", leaderAddr), errs.ZapError(err))
+			select {
+			case <-ctx.Done():
+				log.Info("stop synchronizing with leader due to context canceled")
+				return
+			case <-time.After(retryInterval):
+			}
+			continue
+		}
+		log.Info("server starts to synchronize with leader", zap.String("server", s.name), zap.String("leader", leaderAddr), zap.Uint64("request-index", s.nextSyncIndex))
+		for {
+			resp, err := stream.Recv()
+			failpoint.Inject("syncMetError", func() {
+				err = errors.Errorf("sync met error")
+			})
+			if err != nil {
+				log.Warn("region sync with leader meet error", errs.ZapError(errs.ErrGRPCRecv, err))
+				if err = stream.CloseSend(); err != nil {
+					log.Warn("failed to terminate client stream", errs.ZapError(errs.ErrGRPCCloseSend, err))
+				}
+				select {
+				case <-ctx.Done():
+					log.Info("stop synchronizing with leader due to context canceled",
+						zap.String("server", s.name), zap.Uint64("next-index", s.nextSyncIndex))
+					return
+				case <-time.After(retryInterval):
+				}
+				break
+			}
+			// pd leader expired and needs to reconnect.
+			if e := resp.Header.GetError(); e != nil {
+				log.Warn("server broken the connection, it needs to connect again", zap.String("error-message", e.GetMessage()))
+				s.triggerMembershipCheck()
+				return
+			}
+			if s.nextSyncIndex != resp.GetStartIndex() {
```

Comment:
Under what circumstances does this occur?

Reply:
PD leader has changed or client/server loss some region

Reply:
Yes, we may lose some region changes when the PD leader switches happen. I will design the mismatch index issue later.